### PR TITLE
Enable auth testing for e2e.sh

### PR DESCRIPTION
**Pull request by myidpt**
When auth is enabled, deploy Istio CA (istio/auth), and all proxy-proxy traffic are encrypted by the certs/keys generated by CA.

There are three testing modes, specified by the "auth" flag:
1. enable - Auth is enabled in the test.
2. disable - Auth is disabled in the test.
3. both - Run two rounds of tests, first round with auth disabled, and second round with auth enabled.

Will wait for PR #494 to merge.

### CODE REVIEW

File: docker/prepare_proxy.sh [hunk 1]
```diff
@@ -41,6 +41,13 @@ if [[ -z "${ISTIO_PROXY_PORT-}" ]] || [[ -z "${ISTIO_PROXY_UID-}" ]]; then
 fi
 
 iptables -t nat -A PREROUTING -p tcp -j REDIRECT --to-port ${ISTIO_PROXY_PORT}
+
+# To make sure when pod A wants to talk to service A, which is backed by pod A,
+# the traffic is going through proxy twice, client-side and server-side.
+# lo traffic doesn't go through PREROUTING, so needed to be processed in OUTPUT.
+iptables -t nat -A OUTPUT -p tcp -j REDIRECT -o lo \
+  ! -d 127.0.0.1/32 --to-port ${ISTIO_PROXY_PORT}
```
Comment (kyessenov) [inline docker/prepare_proxy.sh]: remove this and rebase.

File: test/integration/driver.go [hunk 2]
```diff
@@ -79,35 +86,65 @@ var (
...
+		runTest()
+		break
+	case "disable":
+		enableAuth = false
+		runTest()
+		break
+	case "both":
+		enableAuth = false
+		runTest()
+		time.Sleep(30 * time.Second)
```
Comment (kyessenov) [inline test/integration/driver.go]: why do we need sleep here? they run in different namespaces.

File: test/integration/driver.go [hunk 3]
```diff
@@ -79,35 +86,65 @@ var (
...
 	flag.BoolVar(&params.logs, "logs", true, "Validate pod logs (expensive in long-running tests)")
 }
 
 func main() {
 	flag.Parse()
+	switch params.auth {
+	case "enable":
+		enableAuth = true
+		runTest()
+		break
```
Comment (kyessenov) [inline test/integration/driver.go]: no need to break in golang, it's automatic

File: test/integration/driver.go [hunk 4]
```diff
@@ -79,35 +86,65 @@ var (
...
+
+func runTest() {
+	glog.Infof("\n--------------- Run tests with auth: %t ---------------", enableAuth)
+
 	setup()
 
 	for i := 0; i < params.count; i++ {
 		glog.Infof("Test run: %d", i)
-		check((&reachability{}).run())
+		check((&reachability{}).run(enableAuth))
```
Comment (kyessenov) [inline test/integration/driver.go]: put `enableAuth` into `params` and do not pass it here

File: docker/prepare_proxy.sh (hunk 1 above)
Comment (myidpt) [inline docker/prepare_proxy.sh]: Will do it when the other PR is in.

File: test/integration/driver.go (hunk 3 above)
Comment (myidpt) [inline test/integration/driver.go]: Done.

File: docker/prepare_proxy.sh (hunk 1 above)
Comment (kyessenov) [inline docker/prepare_proxy.sh]: It's in

File: test/integration/driver.go (hunk 2 above)
Comment (myidpt) [inline test/integration/driver.go]: Done.

File: test/integration/driver.go (hunk 4 above)
Comment (myidpt) [inline test/integration/driver.go]: Done.

**Comment by kyessenov** [comment]
@sebastienvas jenkins broken again for smoketest. I cannot trigger it in the UI, says build not found.

**Comment by sebastienvas** [comment]
jenkins build manager/e2e-smoketes

**Comment by sebastienvas** [comment]
jenkins build manager/e2e-smoketest

**Comment by sebastienvas** [comment]
jenkins build manager/e2e-smoketest

**Pull request merged**